app/src/main/java/org/mozilla/tv/firefox/architecture/KillswitchLinearLayout.kt

```diff
@@ -0,0 +1,82 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.architecture
+
+import android.content.Context
+import android.util.AttributeSet
+import android.view.View
+import android.widget.LinearLayout
+import org.mozilla.tv.firefox.components.locale.LocaleManager
+import java.util.Locale
+
+sealed class KillswitchLocales {
+    object All : KillswitchLocales()
+    class ActiveIn(vararg val locales: Locale) : KillswitchLocales()
+}
+
+/**
+ * Can be used to wrap some other [View] in order to prevent it from ever being shown if
+ * certain locale and experiment requirements are not met.
+ *
+ * This is done to prevent bugs similar to #2133. In that bug, a view was set to visibility
+ * GONE when disabled, however its animation toggled that visibility back to VISIBLE.
+ * [KillswitchLinearLayout] allows us to be sure that something will not be displayed if
+ * requirements are not set by intercepting [setVisibility] calls.
+ */
+class KillswitchLinearLayout : LinearLayout {
+    constructor(context: Context) : this(context, null)
+    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
+
+    private var desiredVisibility = this.visibility
+
+    private var isAllowedByCurrentExperiment: Boolean? = null
+    private var allowedInLocales: KillswitchLocales? = null
+
+    fun setRequirements(isAllowedByCurrentExperiment: Boolean, allowedInLocales: KillswitchLocales) {
+        this.isAllowedByCurrentExperiment = isAllowedByCurrentExperiment
+        this.allowedInLocales = allowedInLocales
+        visibility = desiredVisibility
+    }
+
+    /**
+     * Sets the visibility to gone if any of the following are true:
+     * 1) Locale/Experiment requirements have not been set
```

Comment:
Hm, I didn't want to enforce settings requirements and visibility in any particular order, so I didn't see that as an error.  It seems like this should get caught at dev time.  If requirements aren't set, the feature just won't ever show up.

Reply:
fwiw, if we did enforce a contract on how this View is used, the View has fewer states, making the code simpler to understand, decreasing surface area for bugs. While it's not a lot of code right now so it's fairly negligible, it could be a bigger deal if the View adds more states later.

I don't have context on how this View is intended to be used to know what my preference is but I just wanted to share some more trade-offs I had thought of.

Reply:
Coming back to this, I'm pretty sure it would crash whenever visibility was set in XML, since that would be applied right after inflation, and we don't have a chance to set requirements until later.